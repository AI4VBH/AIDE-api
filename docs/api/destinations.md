<!--
  ~ Copyright 2022 Guy’s and St Thomas’ NHS Foundation Trust
  ~
  ~ Licensed under the Apache License, Version 2.0 (the "License");
  ~ you may not use this file except in compliance with the License.
  ~ You may obtain a copy of the License at
  ~
  ~ http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing, software
  ~ distributed under the License is distributed on an "AS IS" BASIS,
  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  ~ See the License for the specific language governing permissions and
  ~ limitations under the License.
-->

# Destinations

The `destinations` endpoint provides the following APIs to configure the export destinations for workflow results.

## `GET /destinations`

Returns a list of destinations.

### Parameters

N/A

### Responses

| Code | Description |
|------|-------------|
| `200` | Destinations retrieved successfully. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request POST 'http://localhost:5000/destinations'
```

### Example Response

```json
[
  {
    "port": 104,
    "name": "USEAST",
    "aeTitle": "PACSUSEAST",
    "hostIp": "10.20.3.4"
  }
]
```

---

## `POST /destinations`

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| Body | | See the `IDestination` interface for details. |

### Responses

| Code | Description |
|------|-------------|
| `201` | Destination created successfully. |
| `400` | Validation error. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request POST 'http://localhost:5000/destinations' \
    --header 'Content-Type: application/json' \
    --data-raw '{
        "name": "USEAST",
        "hostIp": "10.20.3.4",
        "port": 104,
        "aeTitle": "PACSUSEAST"
    }'
```

### Example Response

```json
{
  "port": 104,
  "name": "USEAST",
  "aeTitle": "PACSUSEAST",
  "hostIp": "10.20.3.4"
}
```

---

## `PUT /destinations/:name`

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:name` | `string` corresponding to the destination name. |
| Body | | See the `IDestination` interface for details. |


### Responses

| Code | Description |
|------|-------------|
| `200` | DICOM destination updated successfully. |
| `400` | Validation error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with details of the validation errors . |
| `404` | DICOM destination cannot be found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request PUT 'http://localhost:5000/destinations/USEAST' \
    --header 'Content-Type: application/json' \
    --data-raw '{
        "name": "USEAST",
        "hostIp": "10.20.3.4",
        "port": 104,
        "aeTitle": "PACSUSEAST"
    }'
```

### Example Response

```json
{
  "port": 104,
  "name": "USEAST",
  "aeTitle": "PACSUSEAST",
  "hostIp": "10.20.3.4"
}
```

---

## `GET /destinations/echo/:name`

Checks the status of a destination

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:name` | `string` corresponding to the destination name. |

### Responses

| Code | Description |
|------|-------------|
| `200` | Echo successful. |
| `404` | Destination not found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |
| `502` | C-ECHO failure. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/destinations/echo/USEAST'
```