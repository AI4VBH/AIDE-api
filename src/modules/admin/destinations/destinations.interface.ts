export interface IDestination {
  port: number;
  name: string;
  aeTitle: string;
  hostIp: string;
}
