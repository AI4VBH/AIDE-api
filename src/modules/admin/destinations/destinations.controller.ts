import { Body, Controller, Get, Inject, Post } from '@nestjs/common';
import { IDestination } from './destinations.interface';
import { DestinationsService } from './destinations.service';

@Controller('destinations')
export class DestinationsController {
  @Inject(DestinationsService)
  private readonly service: DestinationsService;

  @Get()
  getDestinations() {
    return this.service.getDestinations();
  }

  @Post()
  registerDestination(@Body() destination: IDestination) {
    return this.service.registerDestination(destination);
  }
}
