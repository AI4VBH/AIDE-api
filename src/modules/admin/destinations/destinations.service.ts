import { HttpService } from '@nestjs/axios';
import { Inject, Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { firstValueFrom } from 'rxjs';
import { IDestination } from './destinations.interface';

@Injectable()
export class DestinationsService {
  @Inject(HttpService)
  private readonly httpService: HttpService;

  @Inject(ConfigService)
  private readonly configService: ConfigService;

  async getDestinations(): Promise<IDestination[]> {
    const baseURL = this.configService.get<string>('MIG_API_HOST');

    const response = await firstValueFrom(
      this.httpService.get<IDestination[]>('/config/destination', {
        baseURL,
      }),
    );

    return response.data;
  }

  registerDestination(destination: IDestination) {
    return null;
  }
}
