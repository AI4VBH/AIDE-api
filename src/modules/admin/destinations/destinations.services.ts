import { HttpService } from '@nestjs/axios';
import { Inject } from '@nestjs/common';
import { IDestination } from './destinations.interface';

export class DestinationsService {
  @Inject(HttpService)
  private readonly httpService: HttpService;

  getDestinations() {
    return null;
  }

  registerDestination(destination: IDestination) {
    return null;
  }
}
