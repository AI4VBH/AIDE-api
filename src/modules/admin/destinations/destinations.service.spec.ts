import { DeepMocked, createMock } from '@golevelup/ts-jest';
import { HttpService } from '@nestjs/axios';
import { ConfigService } from '@nestjs/config';
import { Test, TestingModule } from '@nestjs/testing';
import { AxiosInstance } from 'axios';
import { makeObservableForTest } from 'test/utilities/test-make-observable';
import { IDestination } from './destinations.interface';
import { DestinationsService } from './destinations.service';

describe('DestinationsService', () => {
  let configServiceMock: DeepMocked<ConfigService>;
  let axios: DeepMocked<AxiosInstance>;
  let httpService: DeepMocked<HttpService>;
  let destinationsService: DestinationsService;

  beforeEach(async () => {
    configServiceMock = createMock<ConfigService>();
    axios = createMock<AxiosInstance>();
    httpService = createMock<HttpService>({
      axiosRef: axios,
    });

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        DestinationsService,
        { provide: HttpService, useValue: httpService },
        {
          provide: ConfigService,
          useValue: configServiceMock,
        },
      ],
    }).compile();

    destinationsService = module.get<DestinationsService>(DestinationsService);
  });

  it('should be defined', () => {
    expect(destinationsService).toBeDefined();
  });

  describe('getDestinations', () => {
    it('returns expected result', async () => {
      const destinations = [
        {
          port: 104,
          name: 'USEAST',
          aeTitle: 'PACSUSEAST',
          hostIp: '10.20.3.4',
        },
        {
          port: 104,
          name: 'USWEST',
          aeTitle: 'PACSUSWEST',
          hostIp: '10.50.3.4',
        },
      ] as IDestination[];
      axios.get.mockResolvedValue({
        status: 200,
        data: destinations,
      });
      httpService.get.mockReturnValue(makeObservableForTest(axios.get));

      const response = await destinationsService.getDestinations();

      expect(httpService.get).toHaveBeenCalled();
      expect(response).toMatchSnapshot();
      expect(response).toEqual(destinations);
    });
  });
});
