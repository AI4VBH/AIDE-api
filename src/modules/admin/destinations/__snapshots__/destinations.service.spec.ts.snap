// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DestinationsService getDestinations returns expected result 1`] = `
[
  {
    "aeTitle": "PACSUSEAST",
    "hostIp": "10.20.3.4",
    "name": "USEAST",
    "port": 104,
  },
  {
    "aeTitle": "PACSUSWEST",
    "hostIp": "10.50.3.4",
    "name": "USWEST",
    "port": 104,
  },
]
`;
